### PR TITLE
Add condition to closeUrl for block editor to handle checklist referrals

### LINKED ISSUE #35452: FSE: Finishing editing your homepage does not take you back to checklist [2]
If I come to the editor via the checklist, when you hit the close button it should take me straight back to the checklist, not the pages screen.

### PR DESCRIPTION
**Pull request by jeraldjuice**
### Summary

After being referred to the block editor from the checklist, this PR implements directing back to the checklist when the block editor close button is clicked.

backtochecklist

* Overrides close button URL similarly to how the previous implementation handles FSE Template post type close buttons.
* A bit of refactoring to _kill the lets_.

<hr>

### Outstanding and questions

- [x] "View Pages"/"View Posts" tooltip on the close button is a bit wonky when you're not actually going back to Pages/Posts. This will probably need to be overridden in the iframe server. Which leads me to the question...
- [x] Does it make more sense to just completely override the close button for active checklist tasks? Similarly to FSE. Might create a better call to action when working through the checklist and solve the above problem as well.
Screen Shot 2019-08-28 at 11 49 31 AM

- [x] Does the default URL for the close button make more sense to just go to the previous route, as opposed to the Posts/Pages listings? (Assuming that a previous route exists, otherwise default to Posts/Pages.)

^ Above questions answered. Outstanding:

- [x] Move functionality into a selector and utilize the 'actionLog' to look further back in the route stack.
- [x] Add a label to back button for checklist navigation and remove "Back to Post/Page" tooltip.
- [x] Ensure it is possible to navigate back to checklist after editing a template part.

<hr>

### Testing

* Spin up a local Calypso testing environment. You will also have to build the wpcom block editor.
* Open a site with an active checklist.
* Test the individual flows:
    * Test the Checklist-to-block editor flow: Ie, by clicking "Try It" under the "Update your homepage" task in the checklist. Expect that the close/back button in the block editor takes you back to the Checklist.
    * Test the neutral back button flow: Ie, by selecting a post from the Posts page or going directly to the block editor by pasting a URL in the address bar. Expect that the close/black button in the block editor takes you back to the Posts (or Pages) listing page.
    * Test the pre-existing FSE flow: Ie, by going to a page, then selecting "Edit Header". Try the back button from the Template part. Expect that it brings you back to the parent page editor.

Fixes #35452

### CODE REVIEW

**Comment by apeatling** [comment]
> Does it make more sense to just completely override the close button for active checklist tasks? Similarly to FSE. Might create a better call to action when working through the checklist and solve the above problem as well.

I really like the call to action as part of the button.

**Comment by shaunandrews** [comment]
I could see adding the label o the button, but we could probably shorten it just to “< Checklist”. Worth considering how this would affect things on mobile. I’d imagine we’d have to hide the label on small screens.

**Comment by jeraldjuice** [comment]
Thanks, @shaunandrews and @apeatling ! I think that the label is really nice when changing the destination of the back button, so I'll add that to this PR as well.

**Comment by jeraldjuice** [comment]
Okay, so interesting thing I found during testing - there are edge cases when navigating from the checklist that wipe `actionLog` entirely, due to essentially navigating in a way that refreshes the app.

Based on that, I've had to scrap the `actionLog` implementation and I think that using a URL query might be a better way to go. This latest push shows roughly how that will work.

A couple of outstanding things that I'll update in a bit once I'm finished with them:

- [ ] This needs to be generalized for _all_ links to the block editor from the checklist. Short of updating each vertical, I'll do this through a helper function that properly transforms links to the block editor from the checklist.
- [ ] The "Back to Page" URL from the FSE template part needs to properly reflect the checklist referral query. 
- [ ] General refactoring to smooth out the roughness + security enhancements.
- [ ] Add back in the 'Back to Checklist' button transform.

**Comment by gwwar** [comment]
@jeraldjuice before you get too deep do you mind posting a few redirect url examples? There may be some options to mark global state, which should be a bit easier to keep track of since this flow may involve many url updates between editing post_content vs header/footer content

**Comment by jeraldjuice** [comment]
I ended up reverting those changes. I figured out what the wiped actionLog was - a dev environment problem, not a real problem. 🤦‍♂ 

I need to add in the condition to show the checklist label and also hide the "View Posts/Pages" label, which I'll do after lunch. Besides that, this is good to test/review.

File: client/state/selectors/get-editor-close-url.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	const checklistRoute = `/checklist/${ getSiteSlug( state, siteId ) }`;
+
+	// Handle returning to parent editor for full site editing templates
+	if ( 'wp_template' === postType ) {
+		return getGutenbergEditorUrl( state, siteId, fseParentPageId, 'page' );
+	}
+
+	// Checking if we should navigate back to the checklist
+	if ( isLastNonEditorRouteChecklist( state ) ) {
+		return checklistRoute;
```
Comment (Copons) [inline client/state/selectors/get-editor-close-url.js]: We only use `checklistRoute` once.
Couldn't we return its value instead?

```es6
if ( isLastNonEditorRouteChecklist( state ) ) {
  return `/checklist/${ getSiteSlug( state, siteId ) }`;
}
```

**Comment by Copons** [comment]
This tests as advertised and codewise LGTM.

I'd maybe question the need to have a full fledged selector for `isLastNonEditorRouteChecklist`, which imho has a very specific use case, instead of putting it into `getEditorCloseUrl` and call it a day.
On the other hand, this way its unit tests might be more straightforward.

(Speaking of, we don't have any tests for `getEditorCloseUrl`!)

File: apps/wpcom-block-editor/src/common/override-close-btn.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/* global calypsoifyGutenberg */
+
+/**
+ * External dependencies
+ */
+import domReady from '@wordpress/dom-ready';
+import { get } from 'lodash';
+
+domReady( () => {
+	const closeButtonInception = setInterval( function() {
```
Comment (gwwar) [inline apps/wpcom-block-editor/src/common/override-close-btn.js]: Would it be possible to follow up in another PR avoiding the setInterval here? Feels like we should be able to listen to some Gutenberg redux state instead.

File: client/state/selectors/test/is-last-non-editor-route-checklist.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ * Internal dependencies
+ */
+import isLastNonEditorRouteChecklist from 'state/selectors/is-last-non-editor-route-checklist';
+import { ROUTE_SET } from 'state/action-types';
+
+const blockEditorAction = { type: ROUTE_SET, path: '/block-editor/page/1' };
+const checklistAction = { type: ROUTE_SET, path: '/checklist/examplesite.com' };
+const otherAction = { type: ROUTE_SET, path: '/some/other/url' };
+
+describe( 'isLastNonEditorRouteChecklist()', () => {
```
Comment (gwwar) [inline client/state/selectors/test/is-last-non-editor-route-checklist.js]: ❤️ for the tests

File: apps/wpcom-block-editor/src/common/override-close-btn.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+			return;
+		}
+
+		clearInterval( closeButtonInception );
+
+		// Add 'Checklist' label when the editor close button navigates back to checklist
+		if ( get( calypsoifyGutenberg, 'closeUrl', '' ).includes( '/checklist/' ) ) {
+			const checklistLabel = document.createElement( 'span' );
+			checklistLabel.className = 'checklist-label';
+			checklistLabel.innerHTML = 'Checklist';
```
Comment (gwwar) [inline apps/wpcom-block-editor/src/common/override-close-btn.js]: i18n? Maybe drop this part from the PR for now

Comment (gwwar) [reply apps/wpcom-block-editor/src/common/override-close-btn.js]: 🤔 We probably need to update this value the server as part of the global script enqueue in wpcom. For example https://github.com/Automattic/wp-calypso/blob/20265560d69e24c9def3e1aa9995866dc38fef5a/apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php#L180

Though we probably don't need to tie this to FSE

**Pull request merged**